### PR TITLE
KNOX-2602 - Added the enabled flag and related token service API into the tokens' metadata.

### PR DESCRIPTION
**Pull request by smolnar82**
## What changes were proposed in this pull request?

There is a new `KNOXTOKEN` service API to enable/disable a Knox-generated token.
Instead of adding another column in the `KNOX_TOKENS` table, a brand new DB table was introduced where an arbitrary number of token metadata can be stored (identified by the given token's ID and the metadata name).
In our alias-based token state management implementations, this is stored as a new JSON field whereas in the file journal-based implementation the new flag is added before the user name.

## How was this patch tested?

Updated and executed JUnit tests.

Additionally, the following manual test steps were executed:

1. Created a token using the tokengen application:

Screen Shot 2021-05-04 at 9 45 52 PM

```
postgres=# select * from knox_tokens kt, knox_token_metadata ktm WHERE kt.token_id = ktm.token_id;
               token_id               |  issue_time   |  expiration   | max_lifetime  |               token_id               | md_name  |                                 md_value                                  
--------------------------------------+---------------+---------------+---------------+--------------------------------------+----------+---------------------------------------------------------------------------
 1e2f286e-9df1-4123-8d41-e6af523d6923 | 1620155983024 | 1620249883013 | 1620760783024 | 1e2f286e-9df1-4123-8d41-e6af523d6923 | enabled  | true
 1e2f286e-9df1-4123-8d41-e6af523d6923 | 1620155983024 | 1620249883013 | 1620760783024 | 1e2f286e-9df1-4123-8d41-e6af523d6923 | userName | admin
 1e2f286e-9df1-4123-8d41-e6af523d6923 | 1620155983024 | 1620249883013 | 1620760783024 | 1e2f286e-9df1-4123-8d41-e6af523d6923 | comment  | comment with a question mark ? and my favorite cartoon tom & jerry !!! :)
(3 rows)
```

2. Tried to enable an already enabled token
```
$ curl -ku admin:admin-password -d "1e2f286e-9df1-4123-8d41-e6af523d6923" -X POST https://localhost:8443/gateway/sandbox/knoxtoken/api/v1/token/enable
{
  "setEnabledFlag": "false",
  "error": "Token is already enabled"
}
```
3. Disabled the token
```
$ curl -ku admin:admin-password -d "1e2f286e-9df1-4123-8d41-e6af523d6923" -X POST https://localhost:8443/gateway/sandbox/knoxtoken/api/v1/token/disable
{
  "setEnabledFlag": "true",
  "isEnabled": "false"
}
```
```
postgres=# select * from knox_tokens kt, knox_token_metadata ktm WHERE kt.token_id = ktm.token_id AND ktm.md_name = 'enabled';
               token_id               |  issue_time   |  expiration   | max_lifetime  |               token_id               | md_name | md_value 
--------------------------------------+---------------+---------------+---------------+--------------------------------------+---------+----------
 1e2f286e-9df1-4123-8d41-e6af523d6923 | 1620155983024 | 1620249883013 | 1620760783024 | 1e2f286e-9df1-4123-8d41-e6af523d6923 | enabled | false
(1 row)
```
4. Tried to disable an already disabled token
```
$ curl -ku admin:admin-password -d "1e2f286e-9df1-4123-8d41-e6af523d6923" -X POST https://localhost:8443/gateway/sandbox/knoxtoken/api/v1/token/disable
{
  "setEnabledFlag": "false",
  "error": "Token is already disabled"
}
```

5. Enabled the token
```
$ curl -ku admin:admin-password -d "1e2f286e-9df1-4123-8d41-e6af523d6923" -X POST https://localhost:8443/gateway/sandbox/knoxtoken/api/v1/token/enable
{
  "setEnabledFlag": "true",
  "isEnabled": "true"
}
```
```
postgres=# select * from knox_tokens kt, knox_token_metadata ktm WHERE kt.token_id = ktm.token_id AND ktm.md_name = 'enabled';
               token_id               |  issue_time   |  expiration   | max_lifetime  |               token_id               | md_name | md_value 
--------------------------------------+---------------+---------------+---------------+--------------------------------------+---------+----------
 1e2f286e-9df1-4123-8d41-e6af523d6923 | 1620155983024 | 1620249883013 | 1620760783024 | 1e2f286e-9df1-4123-8d41-e6af523d6923 | enabled | true
(1 row)
```

6. Renewed the token
```
$ curl -ku admin:admin-password -d "@token.txt" -X POST https://localhost:8443/gateway/sandbox/knoxtoken/api/v1/token/renew
{
  "renewed": "true",
  "expires": "1620243157437"
}
```
```
postgres=# select * from knox_tokens;
               token_id               |  issue_time   |  expiration   | max_lifetime  
--------------------------------------+---------------+---------------+---------------
 1e2f286e-9df1-4123-8d41-e6af523d6923 | 1620155983024 | 1620243157437 | 1620760783024
(1 row)
```
7. Revoked the token
```
$ curl -ku admin:admin-password -d "@token.txt" -X POST https://localhost:8443/gateway/sandbox/knoxtoken/api/v1/token/revoke
{
  "revoked": "true"
}
```
```
postgres=# select * from knox_tokens kt, knox_token_metadata ktm WHERE kt.token_id = ktm.token_id;
 token_id | issue_time | expiration | max_lifetime | token_id | md_name | md_value 
----------+------------+------------+--------------+----------+---------+----------
(0 rows)
```

### CODE REVIEW

File: gateway-server/src/main/java/org/apache/knox/gateway/services/token/impl/JDBCTokenStateService.java [hunk 1]
```diff
@@ -193,8 +194,17 @@ protected void evictExpiredTokens() {
 
   @Override
   public void addMetadata(String tokenId, TokenMetadata metadata) {
+
     try {
-      final boolean added = tokenDatabase.addMetadata(tokenId, metadata);
+      boolean added = false;
```
Comment (moresandeep) [inline gateway-server/src/main/java/org/apache/knox/gateway/services/token/impl/JDBCTokenStateService.java:199]: Chance of running into concurrency issue here with the flag.

**Review by moresandeep — CHANGES_REQUESTED** [review]
Just minor concurrency issue but looks great

File: gateway-server/src/main/java/org/apache/knox/gateway/services/token/impl/JDBCTokenStateService.java (hunk 1 above)
Comment (smolnar82) [reply gateway-server/src/main/java/org/apache/knox/gateway/services/token/impl/JDBCTokenStateService.java]: Thanks, @moresandeep ! I've just submitted a new PS with the fix.

**Review by moresandeep — APPROVED** [review]
Loog good!

**Pull request merged**